druid/src/widget/scope.rs

```diff
@@ -0,0 +1,278 @@
+    lens: L,
+    phantom_in: PhantomData<In>,
+    phantom_state: PhantomData<State>,
+}
+
+impl<L: Lens<State, In>, In, State> LensScopeTransfer<L, In, State> {
+    pub fn new(lens: L) -> Self {
+        LensScopeTransfer {
+            lens,
+            phantom_in: PhantomData::default(),
+            phantom_state: PhantomData::default(),
+        }
+    }
+}
+
+impl<L: Lens<State, In>, In: Data, State: Data> ScopeTransfer for LensScopeTransfer<L, In, State> {
+    type In = In;
+    type State = State;
+
+    fn read_input(&self, state: &mut State, data: &In) {
+        self.lens.with_mut(state, |inner| {
+            if !inner.same(&data) {
+                *inner = data.clone()
+            }
+        });
+    }
+
+    fn write_back_input(&self, state: &State, data: &mut In) {
+        self.lens.with(state, |inner| {
+            if !inner.same(&data) {
+                *data = inner.clone();
+            }
+        });
+    }
+}
+
+enum ScopeContent<SP: ScopePolicy> {
+    Policy {
+        policy: Option<SP>,
+    },
+    Transfer {
+        state: SP::State,
+        transfer: SP::Transfer,
+    },
+}
+
+/// A widget that allows encapsulation of application state.
+///
+/// This is useful in circumstances where
+/// * A (potentially reusable) widget is composed of a tree of multiple cooperating child widgets
+/// * Those widgets communicate amongst themselves using Druids reactive data mechanisms
+/// * It is undesirable to complicate the surrounding application state with the internal details
+///   of the widget.
+///
+///
+/// Examples include:
+/// * In a tabs widget composed of a tab bar, and a widget switching body, those widgets need to
+///   cooperate on which tab is selected. However not every user of a tabs widget wishes to
+///   encumber their application state with this internal detail - especially as many tabs widgets may
+///   reasonably exist in an involved application.
+/// * In a table/grid widget composed of various internal widgets, many things need to be synchronised.
+///   Scroll position, heading moves, drag operations, sort/filter operations. For many applications
+///   access to this internal data outside of the table widget isn't needed.
+///   For this reason it may be useful to use a Scope to establish private state.
+///
+/// A scope embeds some input state (from its surrounding application or parent scope)
+/// into a larger piece of internal state. This is controlled by a user provided policy.
+///
+/// The ScopePolicy needs to do two things
+/// a) Create a new scope from the initial value of its input,
+/// b) Provide two way synchronisation between the input and the state via a ScopeTransfer
+///
+/// Convenience methods are provided to make a policy from a function and a lens.
+/// It may sometimes be advisable to implement ScopePolicy directly if you need to
+/// mention the type of a Scope.
+///
+/// # Examples
+/// ```
+/// #[derive(Data, Lens)]
+/// struct AppState{
```

Comment:
Minor formatting nits, there seem to be spaces missing; rustfmt apparently doesn't catch issues inside doctests, which seems to me an oversight.

Reply:
I ran rustfmt on the example so it looks standardised now